src/v2/auth.rs

```diff
@@ -35,91 +173,105 @@ impl Client {
             .await?;
 
         trace!("GET '{}' status: {:?}", r.url(), r.status());
-        let a = r
-            .headers()
+        r.headers()
             .get(reqwest::header::WWW_AUTHENTICATE)
-            .ok_or_else(|| Error::from("get_token: missing Auth header"))?;
-        let chal = String::from_utf8(a.as_bytes().to_vec())?;
+            .ok_or_else(|| {
+                Error::from(format!(
+                    "missing {:?} header",
+                    reqwest::header::WWW_AUTHENTICATE
+                ))
+            })
+            .map(ToOwned::to_owned)
+    }
 
-        let (mut auth_ep, service) = parse_hdr_bearer(chal.trim_start_matches("Bearer "))?;
+    /// Perform registry authentication and return the authenticated client.
+    ///
+    /// If Bearer authentication is used the returned client will be authorized for the requested scopes.
+    pub async fn authenticate(mut self, scopes: &[&str]) -> Result<Self> {
+        let credentials = if let Some(credentials) = self.credentials.clone() {
+            credentials
+        } else {
+            bail!("cannot authenticate without credentials");
+        };
 
-        trace!("Token provider: {}", auth_ep);
-        if let Some(sv) = service {
-            auth_ep += &format!("?service={}", sv);
-            trace!("Service identity: {}", sv);
-        }
+        self.auth = None;
```

Comment:
I think is not needed. If it is actually needed, it's a good sign that we should re-arrange this logic because we have unwanted side-effects leaking into the authentication logic from previous state.

Reply:
It's a relic from the earlier code which double-checks that the authentication was successful. I'm confident enough to remove it.